### ISSUE TITLE
Advanced Grid Layout

### ISSUE THREAD

**Issue by sunjay** [ProtoArt/spritec#153]
Currently, we have a very simple layout engine that only supports a single layout algorithm. That algorithm can produce a regular grid where every row and every column is the same size. There is no way to specify alignment within cells, or make different rows and columns different sizes.

It would be great if we could support a more powerful grid layout algorithm. We can take inspiration from [CSS grid layout](https://www.w3.org/TR/css-grid-1/) and add more properties to the layout nodes to help customize the grid.

## Mentoring Instructions

These instructions aren't complete. This is only an idea of what would have to be changed.

https://github.com/ProtoArt/spritec/blob/482cfdfe748c721451582a917977f3ac9cdb8baf/src/renderer/render_node.rs#L13-L29

Here, we should replace `RenderLayout` an enum instead of a struct and add a wrapper over `RenderNode` to specify grid-specific options for each cell:

```rust
/// Lays out one or more nodes in a given configuration
#[derive(Debug)]
pub enum RenderLayout {
    Grid {
        // ...properties for the grid itself (from CSS grid)...

        /// The nodes to layout
        cells: Vec<GridCell>,
    },
}

/// A cell on a grid
#[derive(Debug)]
pub struct GridCell {
    /// ...properties for this cell (from CSS grid)...

    /// The node to render in this cell
    pub node: RenderNode,
}
```

We should then reimplement whatever we need to in `layout.rs` to support the algorithm. You may want to move the implementation to `layout/grid.rs` and add `mod grid;` to `layout.rs` since it's going to be a lot of code.

https://github.com/ProtoArt/spritec/blob/482cfdfe748c721451582a917977f3ac9cdb8baf/src/renderer/layout.rs#L47-L155

# Notes

- [x] works kind of like a [CSS table](https://www.w3.org/TR/CSS2/tables.html#auto-table-layout) (but much simpler)
- [x] ~~input: a list of nodes and a number of columns~~
    - [x] `Empty` is used to fill in any gaps for blank grid cells
    - [x] Any `Empty`s at the end of the list are optional (i.e. the algorithm should work without them)
- [x] another option: instead of passing a number of columns, we can change the input format to be `Vec<Vec<GridCell>>` (where `GridCell` has the column span, row span, and node) -- might be more intuitive
    - [x] ~~idea: should we get rid of empty and store `Option<GridCell>` instead?~~
- [ ] dynamic column width, per column
    - [ ] with option on the grid to set a static column width on all columns? (error if this makes some images not fit)
- [ ] dynamic row height, per row
    - [ ] with option on the grid to set a static column width on all columns? (error if this makes some images not fit)
- [x] column span
- [x] row span
- [x] error if column span or row span pushes part of a cell outside the grid
- [x] need to decide what we should do if cell size ends up being greater than image size
    - [x] image should probably be centered in the cell?
    - [x] ~~maybe image size should be treated as a min-size value and the dimensions we render at should be the cell size?~~

**Comment by sunjay** [comment]
First version of this will be a heavily simplified version focusing on 1) refactoring the existing code and 2) getting rowspan and colspan working.

Features:

- [x] Input (i.e. the new fields of `GridLayout`)
    - [x] Number of rows in grid
    - [x] Number of columns in grid
    - [x] Cell width in px of every cell
    - [x] Cell height in px of every cell
    - [x] `Vec<Vec<RenderNode>>` to represent the cells starting from the top-left corner and going row-wise
        - [x] There may be gaps in the rows created by `Empty` (including at the start of a row)
        - [x] There may be fewer cells in each row than the number of rows in the grid
        - [x] There may be fewer rows than the number of columns in the grid
- [x] colspan - specified per cell
- [x] rowspan - specified per cell
- [x] If the cell size is greater than the image size, the image is centered in the cell
- [x] Algorithm should always output *something* for every cell, even if that is an `Empty` (this will make extending it later easier)
    - Note that rowspan and colspan make a single cell larger, they do not result in multiple cells
- [x] Error if an image can't fit in the cell width or cell height

In summary, this will give us a simple fully user-specified grid with a fixed cell size that a user can choose to put bigger things in using rowspan and colspan.

The user in this case is @sunjay using our code as an external crate in #31.
